### PR TITLE
[WIP] Updated client notification

### PR DESCRIPTION
**Pull request by woogerboy21**
This is a fix / partial fix for #1214 

This PR adds the ability for client update checking based against the server that is being connected to.  How this works is, upon successful login a command is sent from the client to the server with the clients version string.  That string is then compared by the server to its own version string information and calculates if the servers version is newer than the clients version and if depending on the results sends a response back to the client containing the servers version string and true/false for a client upgrade.  The client then runs through its own checks to see if the expected result for update matches and if it does , it notifies the user that an updated client is available.

Both parties have the ability to disable the functionality preventing either update checks from occurring from the client or update true / server version string information from being returned to the client from the server.

The reason for this transaction type checking on both client and server parts is the future expectation for signed installers allowing for a trusted version of the code base.  If the client (or server) knows that they are in fact from a trusted source they can guarantee that there side calculations for if a client check is correct.  If that type of solution never comes to fruition the functionality will still operate, it just is doing more work than needed.

Another bit of functionality has been added to allow additional flexibility.  The client side check is designed to be a fall back type solution to update checking.  There is a variable created in the settings cache (TrustedSourceClientUpdateSuccess) that can be flipped to disable the client update check at login portion of the functionality leaving room for another update solution to occur prior to this version string type commit check to be performed.  With this functionality in place when another (non-version) type update solution comes around it can be worked in very easily and become the primary update check solution utilizing this as a fallback method.

### CODE REVIEW

**Comment by woogerboy21** [comment]
Side note.  I may want to rename that fallback settings cache variable name to something more simple.

**Comment by poixen** [comment]
Next up: Auto update clients. 

Chop chop!

File: cockatrice/src/remoteclient.cpp [hunk 1]
```diff
@@ -348,3 +358,80 @@ void RemoteClient::disconnectFromServer()
...
+                 *  run the same type of date/commit calculations and verify the results are the same.  If
+                 *  there is a change in the logic here, make sure you update the logic in the server_protocolhandler.cpp
+                 *  file to match the clientUpdateAvailable results.
+                 */
+
+
+                QString serverVer = QString::fromStdString(resp.server_version()).simplified();
+                QString clientVer = QString::fromStdString(VERSION_STRING);
+                QDate clientDate = QDate::fromString(serverVer.mid(9, 10), "yyyy'-'MM'-'dd");
+                QDate serverDate = QDate::fromString(clientVer.mid(9, 10), "yyyy'-'MM'-'dd");
```
Comment (ZeldaZach) [inline cockatrice/src/remoteclient.cpp]: Did you get these `QDate`s backwards?

File: cockatrice/src/remoteclient.cpp [hunk 2]
```diff
@@ -249,6 +258,7 @@ void RemoteClient::doConnectToServer(const QString &hostname, unsigned int port,
     userName = _userName;
     password = _password;
     QString clientid = settingsCache->getClientID();
+    QString clientver = (VERSION_STRING);
```
Comment (ZeldaZach) [inline cockatrice/src/remoteclient.cpp]: Are the `()` necessary?

File: cockatrice/src/remoteclient.cpp [hunk 3]
```diff
@@ -348,3 +358,80 @@ void RemoteClient::disconnectFromServer()
...
+                QDate serverDate = QDate::fromString(clientVer.mid(9, 10), "yyyy'-'MM'-'dd");
+                bool clientUpdateAvailable = false;
+
+                if (clientVer.isEmpty()) {
+                    // client version not available OR client does not have code for sending the information
+                    clientUpdateAvailable = true;
+                } else if (serverVer.isEmpty()) {
+                    // server version not available OR server does not have the code for sending the information
+                    clientUpdateAvailable = false;
+                } else if (clientDate.operator==(serverDate)) {
```
Comment (ZeldaZach) [inline cockatrice/src/remoteclient.cpp]: Do we need the `()` around `serverDate`?

**Comment by tooomm** [comment]
Didn't you said once you don't want to leak information about the server?
To prevent people from exploiting known bugs or cheating on unpatched servers?

> That string is then compared by the server to its own version string information and calculates if the servers version is newer than the clients version and if depending on the results **sends a response back to the client containing the servers version string and true/false for a client upgrade**.

**Comment by Daenyth** [comment]
I think this isn't the best approach.

By adding the version string negotiation to the protocol, all possible server implementations need to keep their version strings in sync with servatrice, and if they don't, clients will think the server is outdated and doesn't support new features. All possible client implementations need to keep their version strings in sync with desktop cockatrice, or the server will tell them they are outdated even if they support all current features.

For example, suppose the most recent version of cockatrice is 2015-09-01. I'm running Robotrice for Android, version 2015-08-25. It supports all the same features that cockatrice does, but now the server tells me that my client is out of date.

Instead, I propose that we negotiate features in the protocol, not versions. Versions don't matter. If I make a custom compilation from a branch that merges something new in, my version string will be newer, but my feature set might not be different at all - could just be a UI change.

My proposal is instead to add a new field to the login action. The client will inform the server of the features it supports. The server will check this against its own feature set. If the client is missing features, the server will inform the client which features it is missing and the client can display a warning to the user. If the client supports features the server doesn't, the server will inform the client that those features may not work. For the official cockatrice/servatrice, we can add a constant somewhere in `common/`. It should contain only features that relate to the protocol; UI features, server updates, etc, won't matter here. We could potentially include the set into a protobuf type so that other client/server implementations can use the same api.

Example:
All features
- reveal_bottom_card
- register_accounts
- frobnicate_widgets
- send_client_id

Client -> Server
```
features: { reveal_bottom_card, register_accounts, frobnicate_widgets }
```

Server config:
```
{ reveal_bottom_card: optional
  send_client_id: required,
  register_accounts: optional}
```

Server -> client
```
{ client_missing_required_features: [ send_client_id ]
  server_does_not_support: [ frobnicate_widgets ] }
```

Client displays:
> Login unsuccessful. The server requires clients to support the `send_client_id` feature. Update your client.

User updates and client connects again, this time saying it supports `send_client_id`

Client displays:
> Warning: This server does not support the `frobnicate_widgets` feature.

**Comment by woogerboy21** [comment]
How do you propose we maintain the list of features?

**Comment by woogerboy21** [comment]
@tooomm 
I did originally but later on found that the information is already being sent back to the client upon login.

**Comment by Daenyth** [comment]
When we add new protocol actions, we add a new feature to the list and mark it as either required or optional. The required/optional part could be up to server admin even, it could just be a QSet in common.

File: cockatrice/src/remoteclient.cpp (hunk 1 above)
Comment (woogerboy21) [inline cockatrice/src/remoteclient.cpp]: Yes I did.... it was late.

File: cockatrice/src/remoteclient.cpp (hunk 3 above)
Comment (woogerboy21) [inline cockatrice/src/remoteclient.cpp]: Was following the Qt page for syntax reference.  I could try it without.

**Comment by woogerboy21** [comment]
So you want to pass all the features as a string that has some common delimiter in the protocol right?

**Comment by Daenyth** [comment]
It could be a `repeated string` in the proto

**Pull request closed without merging**